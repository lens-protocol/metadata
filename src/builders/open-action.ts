import { evaluate } from './ValidationError';
import { RecursiveUnbrand } from './utils';
import { StringAttribute } from '../MetadataAttribute';
import { OpenActionMetadataSchema, OpenActionSchemaId, OpenActionMetadata } from '../open-actions';
import { Prettify } from '../utils';

/**
 * @private
 * @privateRemarks MUST stay very @private to produce usable docs
 */
type OpenActionDetails = Prettify<RecursiveUnbrand<Omit<OpenActionMetadata, '$schema'>>>;
/**
 * All {@link OpenActionMetadata} fields without `$schema` a with:
 * - `attributes` defaults to an empty array
 */
export type OpenActionOptions = OpenActionDetails & {
  /**
   * The attributes of the Open Action.
   */
  attributes?: StringAttribute[];
};

/**
 * Creates a valid OpenActionMetadata.
 *
 * @category Compose
 * @param input - Use your IDE suggestions for an enhanced development experience
 *
 * @example
 * ```ts
 * const metadata = openAction({
 *   name: 'My Open Action',
 *   title: 'This is my Open Action',
 *   content: 'Get ready for the future of social interaction!',
 *   authors: ['awesome-dev@lens.xyz'],
 *   initializeABI: [
 *     {
 *       name: 'address',
 *       type: 'address',
 *       indexed: true,
 *     },
 *     {
 *       name: 'price',
 *       type: 'uint256',
 *       indexed: true,
 *     }
 *   ],
 *   processABI: [
 *     {
 *       name: 'collector',
 *       type: 'address',
 *       indexed: true,
 *     }
 *   ]
 * });
 * ```
 */
export function openAction({ attributes = [], ...others }: OpenActionOptions): OpenActionMetadata {
  return evaluate(
    OpenActionMetadataSchema.safeParse({
      $schema: OpenActionSchemaId.LATEST,
      attributes,
      ...others,
    }),
  );
}
