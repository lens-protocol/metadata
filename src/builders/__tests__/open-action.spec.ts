import { describe, expect, it } from '@jest/globals';

import { MetadataAttributeType } from '../../MetadataAttribute.js';
import { openAction } from '../open-action';

describe(`Given the "${openAction.name}" metadata builder`, () => {
  describe(`when using it`, () => {
    it('should return a valid OpenActionMetadata', () => {
      const metadata = openAction({
        name: 'Open Action Name',
        title: 'Open Action Title',
        authors: ['me@lens.xyz'],
        attributes: [
          {
            key: 'twitter',
            type: MetadataAttributeType.STRING,
            value: 'https://twitter.com/johndoe',
          },
          {
            key: 'dob',
            type: MetadataAttributeType.STRING,
            value: '1990-01-01T00:00:00Z',
          },
        ],
        description: `
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        - Donec a diam lectus.
        `,
        initializeABI: [
          {
            type: 'address',
            name: 'address',
            indexed: true,
          },
          {
            type: 'uint256',
            name: 'price',
            indexed: true,
          },
        ],
        processABI: [
          {
            type: 'address',
            name: 'collector',
            indexed: true,
          },
        ],
      });

      expect(metadata).toMatchSnapshot();
    });
  });
});
