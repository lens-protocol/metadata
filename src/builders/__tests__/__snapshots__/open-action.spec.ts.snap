// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the "openAction" metadata builder when using it should return a valid OpenActionMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/open-actions/1.0.0.json",
  "attributes": [
    {
      "key": "twitter",
      "type": "String",
      "value": "https://twitter.com/johndoe",
    },
    {
      "key": "dob",
      "type": "String",
      "value": "1990-01-01T00:00:00Z",
    },
  ],
  "authors": [
    "me@lens.xyz",
  ],
  "description": "
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec a diam lectus. Sed sit amet ipsum mauris.

        - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        - Donec a diam lectus.
        ",
  "initializeABI": [
    {
      "indexed": true,
      "name": "address",
      "type": "address",
    },
    {
      "indexed": true,
      "name": "price",
      "type": "uint256",
    },
  ],
  "name": "Open Action Name",
  "processABI": [
    {
      "indexed": true,
      "name": "collector",
      "type": "address",
    },
  ],
  "title": "Open Action Title",
}
`;
